addons/stock/models/stock_move.py

```diff
@@ -637,22 +639,42 @@ def _generate_serial_numbers(self, next_serial_count=False):
 
         move_lines_commands = []
         location_dest = self.location_dest_id._get_putaway_strategy(self.product_id) or self.location_dest_id
+
+        move_lines = self.env['stock.move.line']
+        # If the picking type uses pre-fill operation, we take already existing move lines.
+        if self.picking_type_id.show_reserved:
+            move_lines = self.move_line_ids.filtered(without_lot_filter)
+        else:
+            move_lines = self.move_line_nosuggest_ids.filtered(without_lot_filter)
         for i in range(0, next_serial_count):
             lot_name = '%s%s%s' % (
                 prefix,
                 str(initial_number + i).zfill(padding),
                 suffix
             )
-            move_lines_commands.append((0, 0, {
-                'lot_name': lot_name,
-                'qty_done': 1,
-                'product_id': self.product_id.id,
-                'product_uom_id': self.product_id.uom_id.id,
-                'location_id': self.location_id.id,
-                'location_dest_id': location_dest.id,
-                'picking_id': self.picking_id.id,
-            }))
-        self.write({'move_line_ids': move_lines_commands})
+            # If we have move lines on which we can write, we edit it...
+            if len(move_lines):
+                move_line = move_lines[0]
+                move_lines -= move_line
+                move_lines_commands.append((1, move_line.id, {
+                    'lot_name': lot_name,
+                    'qty_done': 1,
+                }))
+            # ... otherwise, we create a new move line.
+            else:
+                move_lines_commands.append((0, 0, {
+                    'lot_name': lot_name,
+                    'qty_done': 1,
+                    'product_id': self.product_id.id,
+                    'product_uom_id': self.product_id.uom_id.id,
+                    'location_id': self.location_id.id,
+                    'location_dest_id': location_dest.id,
+                    'picking_id': self.picking_id.id,
+                }))
+        if self.picking_type_id.show_reserved:
```

Comment:
probably not needed self.write({'move_line_ids': move_lines_commands}) should be enough